src/System.Private.Interpreter/src/Internal/Runtime/Interpreter/ILInterpreter.cs

```diff
@@ -2126,5 +2123,321 @@ private void InterpretSwitch(ref ILReader reader, int jmpBase, int[] jmpDelta)
+                case StackValueKind.NativeInt:
+                    length = (int)stackItem.AsNativeInt();
+                    break;
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
+
+            Debug.Assert(length >= 0);
+
+            TypeDesc arrayType = ((TypeDesc)_methodIL.GetObject(token)).MakeArrayType();
+
+            // TODO: Add support for arbitrary  non-primitive types
+            Array array = RuntimeAugments.NewArray(arrayType.GetRuntimeTypeHandle(), length);
+
+            _stack.Push(StackItem.FromObjectRef(array));
+        }
+
+        private void InterpretLoadLength()
+        {
+            Array array = (Array)PopWithValidation().AsObjectRef();
+            _stack.Push(StackItem.FromInt32(array.Length));
+        }
+
+        private void InterpretStoreElement(ILOpcode opcode)
+        {
+            StackItem valueItem = PopWithValidation();
+            StackItem indexItem = PopWithValidation();
+            Array array = (Array)PopWithValidation().AsObjectRef();
+
+            int index = 0;
+            switch (indexItem.Kind)
+            {
+                case StackValueKind.Int32:
+                    index = indexItem.AsInt32();
+                    break;
+                case StackValueKind.NativeInt:
+                    index = (int)indexItem.AsNativeInt();
+                    break;
+                default:
+                    ThrowHelper.ThrowInvalidProgramException();
+                    break;
+            }
+
+            Debug.Assert(index >= 0);
+
+            ref byte start = ref RuntimeAugments.GetRawDataForArray(array);
+            ref byte position = ref Unsafe.Add(ref start, (IntPtr)((nuint)index * RuntimeAugments.GetElementSizeForArray(array)));
+
+            switch (opcode)
+            {
+                case ILOpcode.stelem_i:
+                    Unsafe.Write(ref position, valueItem.AsNativeInt());
+                    break;
+                case ILOpcode.stelem_i1:
+                    Unsafe.Write(ref position, (sbyte)valueItem.AsInt32());
+                    break;
+                case ILOpcode.stelem_i2:
+                    Unsafe.Write(ref position, (short)valueItem.AsInt32());
+                    break;
+                case ILOpcode.stelem_i4:
+                    Unsafe.Write(ref position, valueItem.AsInt32());
+                    break;
+                case ILOpcode.stelem_i8:
+                    Unsafe.Write(ref position, valueItem.AsInt64());
+                    break;
+                case ILOpcode.stelem_r4:
+                    Unsafe.Write(ref position, (float)valueItem.AsDouble());
+                    break;
+                case ILOpcode.stelem_r8:
+                    Unsafe.Write(ref position, valueItem.AsDouble());
+                    break;
+                case ILOpcode.stelem_ref:
+                    Unsafe.Write(ref position, valueItem.AsObjectRef());
```

Comment:
This needs assignability check. It should end up calling `System.Runtime.TypeCast.StelemRef`

Reply:
Either need to expose the method to do this via RuntimeAugments; or cast the array to `object[]` and assign it (`Array.SetValue` does this).